sympy/solvers/solvers.py

```diff
@@ -1275,22 +1275,23 @@ def _has_piecewise(e):
 
 
 def _solve_undetermined(g, symbols, flags):
-    # helper to find independent variable for undetermined coefficients
-    # call and to return a list with a single dictionary else None
-    #                                              a*x + 2*x + b - c
-    symset = set(symbols)                        # {a, b}
-    free = g.free_symbols                        # {a, b, c, x}
-    ex = free - symset                           # {c, x}
-    if len(ex) != 1:
-        ind, dep = g.as_independent(*symbols)    # (2*x - c, a*x + b)
-        ex = ind.free_symbols & dep.free_symbols # {x, c} & {a, x, b} -> {x}
-    if len(ex) != 1:                             # e.g. (a + b)*x + b - c -> {c}, {a, b, x}
-        ex = dep.free_symbols - symset           # {x} = {a, b, x}-{a, b}
-    if len(ex) == 1:
-        ex = ex.pop()
-        sol = solve_undetermined_coeffs(g, symbols, ex, **flags)
-        if type(sol) is dict:
-            return [sol]
+    """solve helper to return a list with one dict (solution) else None
+
+    A direct call to solve_undetermined_coeffs is more flexible and
+    can return both multiple solutions, handle more than one independent
+    variable, and will simplify solutions. Here, we have to be more
+    cautious to keep from solving something that does not look like
+    an undetermined coeffs system -- to minimize the surprise factor.
+    since singularities that cancel are not prohibited in
+    solve_undetermined_coeffs.
+    """
+    # a direct call will simplify, but a call through
+    # solve will not because linear systems are not simplified
+    _flags = dict(flags, simplify=False, dict=True, set=False)
+    if g.free_symbols - set(symbols):
+        sol = solve_undetermined_coeffs(g, symbols, **_flags)
+        if len(sol) == 1:
+            return sol
```

Comment:
```suggestion
    A direct call to solve_undetermined_coeffs is more flexible and
    can return both multiple solutions and handle more than one independent
    variable. Here, we have to be more cautious to keep from solving
    something that does not look like an undetermined coeffs system --
    to minimize the surprise factor since singularities that cancel are not
    prohibited in solve_undetermined_coeffs.
    """
    if g.free_symbols - set(symbols):
        sol = solve_undetermined_coeffs(g, symbols)
        if type(sol) is dict:
            return [sol]
```